### PR TITLE
Add type hints, docstrings, and expand test. Introduced bug fix in _define_variables

### PR DESCRIPTION
**Pull request by Okroshiashvili**
Here are all the type hints and docstrings which included in my previous PR. Moreover, I added some test cases and made some stylistic modifications. Additionally, the PR includes a bug fix for ```_define_variables()``` function and extra test cases to catch possible bugs.

### CODE REVIEW

**Comment by Okroshiashvili** [comment]
@solegalli why 2216479 commit shows as it failed but the CI not?

File: feature_engine/wrappers/wrappers.py [hunk 1]
```diff
@@ -1,18 +1,16 @@
...
-from feature_engine.dataframe_checks import (
-    _is_dataframe,
-    _check_input_matches_training_df,
-)
-from feature_engine.variable_manipulation import (
-    _define_variables,
-    _find_all_variables,
-    _find_numerical_variables,
-)
-
```
Comment (solegalli) [inline feature_engine/wrappers/wrappers.py:14]: modules from feature engine should be ordered at the end of the import, as per pep8. Could we re-order?

**Review by solegalli — CHANGES_REQUESTED** [review]
Thank you so much for such an amount of work! This is really impressive. And I learned a lot from reviewing :)

I like how you use args, returns and raises in the docstrings, and the replacement by the genexp of the listscomps
Thanks a lot!

Could I ask for some general changes:

**1) Code style: black and flake8:**

We are using black and flake8 to keep code style. Would you mind running your scripts by black to unify style?

Should be:

```
from feature_engine.dataframe_checks import (
    _is_dataframe,
    _check_input_matches_training_df,
    _check_contains_na,
)
```

Instead of:
```

from feature_engine.dataframe_checks import (_check_contains_na,
                                             _check_input_matches_training_df,
                                             _is_dataframe)
```

As per black auto-formatting.

2) In all transformers (log, reciprocal, power, boxcox, base, etc) in the fit method, the argument y can be None or a pandas Series. I see that it says Optional string? can you check?

3) Docs: we are using british English, would you mind if we keep words with s instead of z, for example initialise instead of initialize. Not too sure is the best way forward, but we would have to change the entire documentation otherwise. Maybe we do that in a different PR?

4) in the init of almost every transformer (except MathCombination), the parameter variables can take a list of strings, a single string, or None. Optional(List[str]) means that it can take a list of strings. Do I understand this correctly? should it not be Union(List[str], str) = None instead?

5) imports shttps://www.python.org/dev/peps/pep-0008/#imports

Imports should be grouped in the following order:

    Standard library imports.
    Related third party imports.
    Local application/library specific imports.

We should put a blank line between each group of imports.(wich I don't always do :/)

After this, I will make a PR to your repo with some of these changes. Would you mind extending them to the rest of the scripts?

Or tell me otherwise if you disagree. 

Thank you!

**Comment by Okroshiashvili** [comment]
@solegalli I'm again late :) I will consider all the suggestions you made and will adjust my PR. I need some time to do all of these.

File: feature_engine/base_transformers.py [hunk 2]
```diff
@@ -1,22 +1,37 @@
...
 
-from feature_engine.dataframe_checks import (
-    _is_dataframe,
-    _check_input_matches_training_df,
-    _check_contains_na,
-)
+from feature_engine.dataframe_checks import (_check_contains_na,
+                                             _check_input_matches_training_df,
+                                             _is_dataframe)
 from feature_engine.variable_manipulation import _find_numerical_variables
```
[the remark this answers is not in the document]
Comment (Okroshiashvili) [reply feature_engine/base_transformers.py:13]: I used black. Never tried before and looks amazing. Thanks for your suggestion. :)

File: feature_engine/base_transformers.py [hunk 3]
```diff
@@ -1,22 +1,37 @@
...
+        Fits the transformation to the DataFrame.
+
+        Args:
+            X: Pandas DataFrame to fit the transformation
+            y: This parameter exists only for compatibility with
+            sklearn.pipeline.Pipeline.
+            Defaults to None.
+
+        Returns:
+            DataFrame with fitted transformation
```
[the remark this answers is not in the document]
Comment (Okroshiashvili) [reply feature_engine/base_transformers.py:31]: In the initial source the ```fit()``` method returns DataFrame. I'm changing code that the ```fit()``` will return ```self```

File: feature_engine/transformation/boxcox.py [hunk 4]
```diff
@@ -38,24 +40,28 @@ class BoxCoxTransformer(BaseNumericalTransformer):
         transfomration} pairs. These are determined automatically.
     """
 
-    def __init__(self, variables=None):
+    def __init__(self, variables: Optional[List[str]] = None) -> None:
```
[the remark this answers is not in the document]
Comment (Okroshiashvili) [reply feature_engine/transformation/boxcox.py:44]: But here the existing None makes it difficult to only pass string or list of strings

**Comment by solegalli** [comment]
@Okroshiashvili all good, merged! Thanks a lot!

**Pull request merged**